Revamp module initialization

djmitche opened the pull request:
The diff is pretty formidable, but the individual commits are each pretty bite-sized and tell a nice story.

The overall idea is this: each module registers a `Moduler`, and that `Moduler` gives access to a `Runner` via its `NewRun` method.  In principle, modules could store module-global data (say, a timestamped cache of all installed packages for the `pkg` module) in the `Moduler` object, and should store run-specific data (counts of events, etc.) in the `Runner` object, rather than in global variables :wink:.

jvehent commented:
Tests are failing on `src/mig/testutil/modtest.go:15: execution declared and not used`

djmitche commented:
Sorry it took me so long to loop back to this.. and I don't know how I pushed a change with that bit missing, but hey -- this runs 'make' successfully for me now.

djmitche commented:
ping?

jvehent commented:
Ohai from Canterbury, UK. I'm traveling for another couple of weeks, so it may take a while to get eyes on this PR unfortunately :(

On June 4, 2015 6:19:59 PM GMT+01:00, "Dustin J. Mitchell" <notifications@github.com> wrote:
>ping?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/mozilla/mig/pull/64#issuecomment-108980754

-- 
sent from mobile

djmitche commented:
Oh, right, I totally forgot!

No worries then.  I've got another Go project to tide me over (
https://github.com/djmitche/relengapi-proxy)

Dustin

On Thu, Jun 4, 2015 at 5:10 PM, Julien Vehent <notifications@github.com>
wrote:

> Ohai from Canterbury, UK. I'm traveling for another couple of weeks, so it
> may take a while to get eyes on this PR unfortunately :(
>
> On June 4, 2015 6:19:59 PM GMT+01:00, "Dustin J. Mitchell" <
> notifications@github.com> wrote:
> >ping?
> >
> >---
> >Reply to this email directly or view it on GitHub:
> >https://github.com/mozilla/mig/pull/64#issuecomment-108980754
>
> --
> sent from mobile
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/mig/pull/64#issuecomment-109053019>.
>

jvehent commented on doc/modules.rst:
```diff
@@ -30,20 +30,37 @@ Registration
 A module registers itself at runtime via its ``init()`` function which must
-call ``modules.Register`` with a module name and a "runner function" returning
-an anonymous object (``interface {}``) that must implement the
+call ``modules.Register`` with a module name and an object implementing
```
go doesn't have objects. `modules.Moduler` is an interface. Could you replace `object` with `interface` in that sentence?

djmitche replied:
Well, interfaces don't implement things.  At least in C, an instance of a struct is idiomatically referred to as an object.  How about "value"

djmitche replied:
Actually, [go does have objects](https://golang.org/doc/faq#Is_Go_an_object-oriented_language).  It's not endowed with all of the behaviors a Java or C++ programmer might expect -- just like a C object -- but the term is still valid

djmitche replied:
Ah, how about "instance"?

jvehent commented on doc/modules.rst.html:
```diff
@@ -527,4 +527,4 @@
 </body>
-</html>
\ No newline at end of file
+</html>
```
note to self: this needs to be regenerated with the new content.

jvehent commented on src/mig/agent/agent.go:
```diff
@@ -163,8 +163,8 @@ func runModuleDirectly(mode string, args []byte, pretty bool) (out string) {
 	// instanciate and call module
-	modRunner := modules.Available[mode].Runner()
-	out = modRunner.(modules.Moduler).Run(os.Stdin)
+	run := modules.Available[mode].NewRun()
```
If we're going to rely on `NewRun()` being present, it needs to be part of the `modules.Runner` interface to break compilation if a module forgets to implement it.

```go
type Runner interface {
 	Run(io.Reader) string
 	NewRun() interface{}
 	ValidateParameters() error
 }
```

jvehent replied:
scratch that, it's implemented in the `modules.Moduler` interface.
(really wish github could stage comments...)

djmitche replied:
Yeah, the name "Moduler" seemed appropriate for values representing a module, where it used to name values representing a run of a module.

jvehent commented on src/mig/client/cmd/main.go:
```diff
@@ -71,7 +71,7 @@ func main() {
 		migrc, show, render, target, expiration, afile string
 		modargs                                        []string
-		modRunner                                      interface{}
+		run                                      interface{}
```
indent (gofmt)

jvehent commented:
It looks good. I'll play a bit more with the interfaces and implement a module that persists data between runs. 
Could you rebase against master? @ameihm0912 added a few tests to the `pkg` module that now break in your branch.
`src/mig/modules/pkg/pkg_test.go:25: undefined: Runner`

djmitche commented:
rebased and fixed.

I'm not sure why travis is failing; it looks like a bug in one of the dependencies:
```
GOPATH=/home/travis/gopath/src/github.com/mozilla/mig GOOS=linux GOARCH=amd64 go get -u golang.org/x/net/icmp
package golang.org/x/net/icmp
	imports golang.org/x/net/internal/iana
	imports golang.org/x/net/internal/iana
	imports golang.org/x/net/internal/iana: use of internal package not allowed
```

jvehent commented:
Looks like a bug in the `tip` of go. I've updated travis build conf in #69 to use `go 1.4.2` and it passed without error, so if you rebase again we should be good.

djmitche commented:
re-pushed

djmitche commented:
Anything else before this gets merged?  It's a high risk of bitrotting..

djmitche commented:
http://www.gifbin.com/bin/062011/1307380901_time-lapse-strawberry-rotting.gif

jvehent commented:
Stay put, I'm driving through wonderland!
![bc](https://scontent-sea1-1.xx.fbcdn.net/hphotos-xaf1/t31.0-8/10914913_10153022214853391_8437529620638171525_o.jpg)

jvehent commented:
I finished the review today and reread the updated doc. It all looks good. Merging it.
Thank you Dustin! That's a great contribution. And sorry that it took a while to merge it ;)
